dashboard/test/lib/services/registration_reminder_test.rb

```diff
@@ -14,7 +14,7 @@ class Services::RegistrationReminderTest < ActiveSupport::TestCase
     Timecop.freeze do
       # Initial creation: No reminders
       application_hash = build :pd_teacher_application_hash, regional_partner_id: create(:regional_partner).id
-      application = create :pd_teacher_application, form_data_hash: application_hash
+      application = create :pd_teacher_application, form_data_hash: application_hash, pd_workshop_id: (create :summer_workshop).id
```

Comment:
Any idea why a workshop is needed here but not down below? What's triggering the read of the accepted template?

Reply:
I believe the lower tests are just checking for when a reminder should be sent (i.e. within 2 weeks vs. over 2 weeks) but not actually producing the email to be sent, whereas the first test is actually attempting to test the emails. And in this case, `registration_reminder.html.haml` requires a `workshop.id` on the first line.

Reply:
Ah yes, that'd do it!

But now I have another question 😅 . It looks like we're stubbing out deliver_email
```
  setup_all do
    Pd::Application::TeacherApplication.any_instance.stubs(:deliver_email)
  end
```

but if it's hitting that template, that stub isn't doing anything. I'm curious if we need more specific stubbing?

Reply:
Good call! I tried testing with nothing in the setup_all and it all still passed so I'm wondering if it's necessary in this file?

Reply:
Ah interesting. Hmm 🤔 . I think it's worth it to remove but not investigate further. I'll go ahead and do this since we're wanting to ensure this gets merged in before DTP today.

Reply:
Confirmed that with
```
setup do
    Pd::Application::TeacherApplication.any_instance.stubs(:deliver_email)
  end
```
we get passing tests, even without the pd_workshop id.

```
  setup_all do
    Pd::Application::TeacherApplication.any_instance.stubs(:deliver_email)
  end
```
gives an error on the template, only without pd_workshop_id.

Trying to setup with
```
    Pd::Application::TeacherApplicationMailer.stubs(:registration_reminder).returns(
      mock {|mail| mail.stubs(:deliver_now)}
    )
```
gives an error using `setup_all` but no error on `setup`, both without pd_workshop_id.

I'm thinking since we don't actually want to run `deliver_now` (which happens in `deliver_email`), stubbing the `deliver_email` makes sense. Does that sound okay?

Perhaps there's a larger question of "Do we want to test the templates in these ruby tests?" If the answer is yes, we should figure out how to stub the sending of the email while still using the template. But I also wonder if the mailer previews serve as a good test of the templates.

Reply:
Okay I updated both tests with the knowledge that the stubbing isn't working on all tests with `setup_all`. Mind taking a look and giving verbal approval before one of us merges?

Reply:
> Perhaps there's a larger question of "Do we want to test the templates in these ruby tests?" If the answer is yes, we should figure out how to stub the sending of the email while still using the template. But I also wonder if the mailer previews serve as a good test of the templates.

it'd be really nice to have automated test coverage of the workshop mailers. the mailer previews are nice for manual testing, but are they being covered by automated tests? either path to automated test coverage seems good to me, but I am wondering if unit tests might be a better path than the mailer previews since the previews serve a slightly different purpose.

Reply:
I think we do have that –– TraverseMailPreviewsTest goes through all previews with filename `./test/mailers/previews/*_preview.rb`. So PdTeacherApplicationMailerPreview goes through there, which is also where we render the https://staging-studio.code.org/rails/mailers/pd_teacher_application_mailer page on staging.

The "missing" link is that we haven't ensured all methods defined in TeacherApplicationMailer are actually rendered in PdTeacherApplicationMailerPreview, although I suppose that's true of all of our tests –– we don't ensure all methods have coverage in a test.

I'm seeing a slightly different pattern in TeacherMailerTest which tests more granularly the templates than TeacherMailerPreview –– but I'm not sure that is necessary.

Reply:
thanks for the context. what is your sense of how well the mailer previews do at providing test coverage? since they are being tested automatically, then it seems fine to count the mailer previews as test coverage.

Reply:
I think the only thing that's missing from the preview tests is that they assume all the information needed is given, as the `build_application` method defines school info, user, status, RP, and workshop. It's unclear what happens if, say, the workshop is undefined. But there are other checks for that on the frontend and the backend ("workshop_present_if_required_for_status"). There is coverage for with and without partner.

I'm curious what your thoughts are on if this kind of coverage is enough –– I don't have a clear grasp on which things should be tested where

Reply:
my thoughts are generally:
1. number one goal is to make sure we don't have any completely broken email templates. given the sad state of honeybadger, we might not notice these right away if one was broken. generally speaking, either type of test coverage seems good to ensure this. so, we are not in an emergency situation, but could still really use a north star.
2. unit (controller) tests really are better for testing these codepaths because
    a. they cover more of the codepath than just the mailer
    b. easier to test multiple combinations 
    c. avoids cluttering up mailer preview with unit tests when you do (b)
    d. focuses on codepaths we can actually hit (e.g. just setting workshop to nil in a mailer preview might not be relevant if there is no way codepath which hits that scenario)
3. we should still keep the mailer preview test coverage since this is "free" test coverage

also this reminds me a lot of the question of storybooks as test coverage. when faced with the option of updating or removing many storybooks, we decided to remove storybooks whose only purpose was to provide test coverage. in our situation we aren't faced with the same choice, but I think it helps support the direction we'd want to go in if push comes to shove.

I'm not sure how cluttered the mailer preview page feels to folks already, so it might not be worth refactoring existing tests to this pattern... however if we want to move in the direction of increasing test coverage on our mailers (which I think we do) then I'd suggest adding unit tests for all the reasons above.